### PR TITLE
TRT-2292: Add support for test isolation

### PR DESCRIPTION
**Pull request by xueqzhan**
This PR adds a testScheduler that 
1. Add conflict handler that will prevent tests with same conflict from running at the same time.
2. Conflicts are detected within conflictGroup, which is used to support future conflict modes: instance, bucket and exec
3. This also adds support for taint/toleration. The idea is to allow some tests running while forbidding all others from running at the same time. 

If taint/toleration function looks ok, I will add the structure in the extension library and adjust here accordingly. 

/hold for review and further test

### CODE REVIEW

File: pkg/test/ginkgo/queue.go [hunk 1]
```diff
@@ -18,6 +20,164 @@ type parallelByFileTestQueue struct {
...
+func getTestBucket(test *testCase) string {
+	// TODO: Implement actual bucket assignment
+	return "bucket-a"
+}
+
+// getTestConflictGroup returns the conflict group for a test.
+// Conflicts are only checked within the same conflict group.
+// The group is determined by the test's isolation mode:
+//   - "instance": conflicts scoped to the running instance
+//   - "bucket": conflicts scoped to a test bucket
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go]: Designing it supporting additional isolation modes is great, but I don't think we should have placeholders for things that aren't implemented. I don't expect to implement `instance` or `bucket` any time soon.

File: pkg/test/ginkgo/queue.go [hunk 2]
```diff
@@ -61,27 +221,51 @@ func queueAllTests(remainingParallelTests chan *testCase, tests []*testCase) {
...
+
+					select {
+					case remainingTests <- t:
+						// Successfully put back in queue
+					case <-ctx.Done():
+						// Context cancelled, decrement pending count since test won't complete
+						decrementAndCloseIfDone(pendingTestCount, remainingTests)
+					}
+				}(test)
+			}
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go]: Requeueing adds test to the end, which I think breaks some concepts @jupierce is working on where we front-load long duration tests to avoid the "long tail" problem.

I would suggest modeling this as a priority queue, and set it up so we always dequeue the next longest test that works with the currently running set, without changing the test's order in the list.

As we don't have the data yet, the priority queue should just start with all the same priorities for all tests.

Comment (stbenjam) [reply pkg/test/ginkgo/queue.go]: Just to clarify, the channels producer/consumer model will have to be different to do this.  You don't have to be bound to this model where workers all consume tests from a channel.  In fact, that is sort of limiting you here.

I think it would be better overall to separate scheduling from the actual running of the test, letting you encapsulate all the scheduling logic into something whose inputs are "currently running tests" and outputs "next test to run" You'll need to handle the case where the test queue is not empty, but we can't run anything yet (efficient blocking), but otherwise this kind of design lets us experiment with additional scheduler options later (like front loading long tests).

I think if you describe this well to Cursor it would produce something reasonable to build on.

Comment (xueqzhan) [reply pkg/test/ginkgo/queue.go]: Ah, just saw this. I have another working version. But let me rework it a little. 

Yeah, Cursor can help a lot to begin. But that guy is really not the best programmer, especially when dealing with racing conditions. :)

File: pkg/test/ginkgo/queue.go [hunk 3]
```diff
@@ -18,67 +17,95 @@ type parallelByFileTestQueue struct {
...
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	// Scan from beginning to find first runnable test
+	for i, test := range ts.tests {
+		conflictGroup := getTestConflictGroup(test)
+
+		// Ensure the conflict group map exists
+		if ts.runningConflicts[conflictGroup] == nil {
+			ts.runningConflicts[conflictGroup] = make(map[string]bool)
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go]: Nit, use a proper `Set` type (k8s has one we use elsewhere here)

File: pkg/test/ginkgo/queue.go [hunk 4]
```diff
@@ -54,34 +192,38 @@ func abortOnFailure(parentContext context.Context) (testAbortFunc, context.Conte
...
+			if scheduler.isEmpty() {
+				// Queue is empty, we're done
 				return
 			}
-			// if the context is finished, simply return
-			if ctx.Err() != nil {
+			// Queue has tests but none can run right now - wait a bit before retrying
+			select {
+			case <-time.After(10 * time.Millisecond):
+				continue
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go]: If we get blocked (maybe a test enforces itself as serial by taints not tolerated by anyone else), we'll have 29 goroutines polling every 10ms.

`sync.Cond()`  would be better

File: pkg/test/ginkgo/queue.go [hunk 5]
```diff
@@ -16,7 +17,144 @@ type parallelByFileTestQueue struct {
...
+			// 3. Remove test from queue
+			ts.tests = append(ts.tests[:i], ts.tests[i+1:]...)
+
+			// 4. Return the test (now safe to run)
+			return test
+		}
+	}
+
+	// No runnable test found
+	return nil
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go]: What if you did this instead?

```golang
for {
     for i, test := range ts.tests {
                ... existing code ...
     }
      
      // No runnable test found
      if len(ts.tests) == 0 {
             return nil // Queue is empty - done
      }

      // Tests exist but all are blocked - wait for signal
      // This releases the lock, sleeps, and re-acquires when woken
      ts.cond.Wait()

     // When we're woken up we loop around again
}
```

Then in MarkTestComplete, you wake everyone up at the end

```
        // Wake up ALL waiting workers - some tests may now be runnable
        ts.cond.Broadcast()
```


This avoids the 10ms busy waiting the workers do

File: pkg/test/ginkgo/queue.go [hunk 6]
```diff
@@ -16,7 +17,144 @@ type parallelByFileTestQueue struct {
...
-type TestFunc func(ctx context.Context, test *testCase)
+// getTestConflictGroup returns the conflict group for a test.
+// Conflicts are only checked within the same conflict group.
+func getTestConflictGroup(test *testCase) string {
+	return "default"
+}
+
+// testScheduler manages test scheduling based on conflicts, taints, and tolerations
+// It maintains an ordered queue of tests and provides thread-safe scheduling operations
+type testScheduler struct {
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go:47]: Do you think we could make a Scheduler interface? So we could potentially have different implementation later.  It also sort of makes the boundaries of the scheduler's responsibilities clearer

**Review by stbenjam — COMMENTED** [review]
Looks really good, just a couple comments

File: pkg/test/ginkgo/test_suite.go [hunk 7]
```diff
@@ -60,6 +62,9 @@ type testCase struct {
 	// suite timeout
 	testTimeout time.Duration
 
+	// isolation defines conflict groups, mode, taints, and tolerations for test isolation
+	isolation extensiontests.Isolation
```
Comment (stbenjam) [inline pkg/test/ginkgo/test_suite.go]: Do we need to copy this or could we just access it via spec?

File: pkg/test/ginkgo/queue.go [hunk 8]
```diff
@@ -16,7 +18,165 @@ type parallelByFileTestQueue struct {
 	commandContext *commandContext
 }
 
-type TestFunc func(ctx context.Context, test *testCase)
+// getTestConflictGroup returns the conflict group for a test.
+// Conflicts are only checked within the same conflict group.
+func getTestConflictGroup(test *testCase) string {
+	return "default"
```
Comment (stbenjam) [inline pkg/test/ginkgo/queue.go:27]: This needs to get replaced from the content in the spec right?

File: pkg/test/ginkgo/test_suite.go (hunk 7 above)
Comment (xueqzhan) [reply pkg/test/ginkgo/test_suite.go]: Doesn't seem that we are copying spec in extensionTestSpecsToOriginTestCases. I can and use the spec instead if that is preferred.

File: pkg/test/ginkgo/queue.go (hunk 8 above)
Comment (xueqzhan) [reply pkg/test/ginkgo/queue.go:27]: spec doesn't define group. Group is designed to support mode.

File: pkg/test/ginkgo/test_suite.go (hunk 7 above)
Comment (stbenjam) [reply pkg/test/ginkgo/test_suite.go]: We definitely should, we have a field for it.  Retry copies it -- looks like maybe an oversight that it wasn't copied in the original creation of them

File: pkg/test/ginkgo/queue.go (hunk 8 above)
Comment (stbenjam) [reply pkg/test/ginkgo/queue.go:27]: Sorry I'm not quite understanding what this function does.  `Isolation` has two fields relevant, `Conflicts` and `Mode`.  I think we agreed to only support `mode=exec` for now, but shouldn't we get the `Conflicts` out of the `Isolation` struct?

Why does this always return `default`?

Comment (xueqzhan) [reply pkg/test/ginkgo/queue.go:27]: The idea was to have a framework to support mode in the future. So conflict is only checked within a conflictGroup. Right now all tests belong to default group and therefore work like mode=exec. But just in case another mode is needed, more conflictGroup will be created for that purpose.

Comment (stbenjam) [reply pkg/test/ginkgo/queue.go:27]: Is there a reason not to call it mode here instead of conflict group, and use the name we're implementing ("exec")? It's not clear to me "group" is linked to the "mode"

Comment (xueqzhan) [reply pkg/test/ginkgo/queue.go:27]: Mode is not clear to indicate a grouping mechanism. But conflictGroup is clear about its functionality of grouping. But configGroup can be used to implement mode. So I think conflictGroup is still better.

**Comment by stbenjam** [comment]
/hold cancel

Just a note: I think this should go to openshift-tests-extension and have origin make use of it instead.  There's a version of `run-suite` in OTE when using standalone binaries, and I'd like to avoid the duplication of code.

https://github.com/openshift-eng/openshift-tests-extension/pull/51 ports this there.  Migrating origin will be a bigger task to make use of it, so we can let this one land for now.
